Add signals for user's discussion activity.

peter-fogg opened the pull request:
[TNL-2497](https://openedx.atlassian.net/browse/TNL-2497)

@dianakhuang @cahrens, can you review?

peter-fogg commented on common/djangoapps/django_comment_common/signals.py:
```diff
@@ -0,0 +1,12 @@
+from django.dispatch import Signal
+
+
+thread_created = Signal(providing_args=['user'])
```
Note: if used in the future, these signals will most likely need to be extended to include more information such as the thread ID. I debated adding that information right now, but decided against it. It seems pointless to speculatively add arguments without knowing if or why they'll be used. The only thing that's needed for teams is the user, and more information can be added without needing to modify the existing code.

peter-fogg commented on lms/djangoapps/django_comment_client/base/tests.py:
```diff
@@ -352,10 +369,10 @@ def inner(self, default_store, module_count, mongo_calls, sql_queries, *args, **
-        (ModuleStoreEnum.Type.mongo, 20, 4, 22),
-        (ModuleStoreEnum.Type.split, 3, 13, 22),
-        (ModuleStoreEnum.Type.split, 20, 13, 22),
+        (ModuleStoreEnum.Type.mongo, 3, 4, 23),
```
Updating `last_activity_at` costs us an extra database query in each of these tests.

cahrens replied:
Really it is looking up the user that is causing the extra query, correct? We don't update last_activity_at unless the user is in a team.

peter-fogg commented on lms/djangoapps/django_comment_client/base/tests.py:
```diff
@@ -866,6 +911,42 @@ def un_flag_comment(self, mock_request, is_closed):
+    def test_endorse_comment(self, mock_request):
+        self._setup_mock_request(mock_request)
+        self.setup_signal_handler(comment_endorsed)
+        self.client.login(username=self.moderator.username, password=self.password)
```
Comments can only be endorsed by discussion staff/moderators/TAs, or the student who posted the question. Since these tests are creating dummy questions, it's easier to test with a moderator.

peter-fogg commented on lms/djangoapps/teams/tests/test_models.py:
```diff
@@ -73,3 +85,50 @@ def test_user_in_team_for_course(self, username, course_id, expected_value):
+        )
+    )
+    @ddt.unpack
+    def test_signals(self, signal, (user, should_update)):
```
Note: there's probably some way to make this DDT data look nicer without needing to use tuple unpacking in the argument list, but I'm not sure exactly how.

dianakhuang commented on common/lib/xmodule/xmodule/modulestore/tests/utils.py:
```diff
@@ -197,3 +198,40 @@ def assert_xss(self, response, xss_content):
         self.assertNotContains(response, xss_content)
+
+
+class MockSignalHandlerMixin(object):
```
I really like the idea that we have a Mixin that will handle signals, but since we're using django signals, I'm not sure if makes sense to put this in `common/lib/xmodule`.

peter-fogg replied:
I'm also not a fan of the placement, but I couldn't figure out where a better place would be. Do you have any suggestions?

cahrens replied:
How about creating a new common/test/utils.py?

cahrens commented on common/djangoapps/django_comment_common/signals.py:
```diff
@@ -0,0 +1,12 @@
+thread_created = Signal(providing_args=['user'])
+thread_edited = Signal(providing_args=['user'])
+thread_voted = Signal(providing_args=['user'])
+comment_created = Signal(providing_args=['user'])
```
I wonder if we really want to differentiate between comments and threads. I suppose it gives more flexibility for the future, but it means registering for a lot of different things.

cahrens replied:
I think I've come around to all the individual signals-- it does make the code much more self-documenting and is better for testing.

cahrens commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -503,6 +511,10 @@ def _do_extra_actions(api_content, cc_content, request_fields, actions_form, con
                 if form_value:
                     context["cc_requester"].vote(cc_content, "up")
+                    # TODO: as noted in django_comment_client, is
+                    # unvoting counted as activity?
```
I would think unvoting would count.

cahrens commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -634,6 +647,7 @@ def update_thread(request, thread_id, update_data):
     # Only save thread object if some of the edited fields are in the thread data, not extra actions
     if set(update_data) - set(actions_form.fields):
         serializer.save()
+        thread_edited.send(sender=None, user=request.user)
```
As we discussed, privileged discussion users can edit other people's threads/comments. I think we need to send both who made the change, and who the thread/comment belongs to.

This is the list of things that privileged users can do--

"edit_content", "delete_thread", "openclose_thread", "endorse_comment", "delete_comment", "see_all_cohorts"

dianakhuang replied:
Maybe a simpler thing would be to only count modifications/signals sent by members of the team? That kind of filtering could happen on the model, and we wouldn't have to change the format of the signal as much.

cahrens commented on lms/djangoapps/teams/models.py:
```diff
@@ -134,3 +160,19 @@ def user_in_team_for_course(cls, user, course_id):
+
+    @classmethod
+    def update_last_activity(cls, user):
+        """Set the `last_activity_at` for both this user and their team."""
```
Should we only do this code conditionally if teams are enabled?

cahrens commented on lms/djangoapps/teams/models.py:
```diff
@@ -134,3 +160,19 @@ def user_in_team_for_course(cls, user, course_id):
+    def update_last_activity(cls, user):
+        """Set the `last_activity_at` for both this user and their team."""
+        try:
+            membership = cls.objects.get(user=user)
```
There can be multiple memberships for the same user (both across courses and when a course allows more than one team per course). We need to send more information about what is being edited.

dianakhuang commented on lms/djangoapps/django_comment_client/base/tests.py:
```diff
@@ -508,6 +536,21 @@ def test_update_thread_wrong_commentable_id(self, mock_get_discussion_id_map, mo
+        response = self.client.post(
+            reverse(
+                "create_comment",
+                kwargs={"course_id": self.course_id.to_deprecated_string(), "thread_id": "dummy"}
```
I know you probably got it from somewhere else in this file, but the correct way to serialize a `CourseKey` is to write `unicode(self.course_id)`.

dianakhuang commented on lms/djangoapps/django_comment_client/base/tests.py:
```diff
@@ -866,6 +911,42 @@ def un_flag_comment(self, mock_request, is_closed):
+        response = self.client.post(
+            reverse(
+                view_name,
+                kwargs={item_id: 'dummy', 'course_id': self.course_id.to_deprecated_string()}
```
`unicode(self.course_id)`

dianakhuang commented on lms/djangoapps/django_comment_client/base/tests.py:
```diff
@@ -866,6 +911,42 @@ def un_flag_comment(self, mock_request, is_closed):
+        response = self.client.post(
+            reverse(
+                'endorse_comment',
+                kwargs={'comment_id': 'dummy', 'course_id': self.course_id.to_deprecated_string()}
```
`unicode(self.course_id)`

peter-fogg commented:
jenkins run a11y, por favor

peter-fogg commented:
FYI @cahrens @dianakhuang, this is ready for re-review. The major changes from last time are:
* Update `last_activity_at` on deletion
* Pass a thread or comment along with a signal as the `post` parameter
* Improve the `MockSignalHandlerMixin` implementation
* Fixed sending `thread_voted` instead of `comment_voted` and vice versa in `discussion_api`
* Only update `last_activity_at` for discussion activity with the standalone context, and if the user performing the activity owns the original post
* Update `last_activity_at` by course

I also went ahead and removed `to_deprecated_string` from everything in the `django_comment_client` tests for consistency.

peter-fogg commented:
jenkins run bokchoy, when you get a chance

cahrens commented on lms/djangoapps/teams/models.py:
```diff
@@ -134,3 +200,22 @@ def user_in_team_for_course(cls, user, course_id):
+        this course.
+        """
+        try:
+            membership = cls.objects.get(user=user, team__course_id=course_id)
```
I still think you need to track the commentable associated with the change. For the MVP, a student can only be in a single team per course. But we plan to relax that assumption, so we shouldn't make this code dependent on it. If the student is in more than one team in the course, there may be multiple memberships. We need the commentable to see which team the user was "active" in.

peter-fogg commented:
@cahrens I've fixed the failing tests, and now updates of `last_activity_at` are filtered by discussion ID, which is unique per-team.

cahrens commented on lms/djangoapps/discussion_api/tests/test_api.py:
```diff
@@ -1328,7 +1330,14 @@ def assert_page_correct(page, page_size, expected_start, expected_stop, expected
 @ddt.ddt
-class CreateThreadTest(CommentsServiceMockMixin, UrlResetMixin, SharedModuleStoreTestCase):
+@disable_signal(api, 'thread_created')
+@disable_signal(api, 'thread_voted')
```
Add coverage for voted too.

peter-fogg replied:
Done.

cahrens commented on lms/djangoapps/discussion_api/api.py:
```diff
@@ -31,6 +31,16 @@
+    comment_edited,
+    comment_voted,
+    comment_deleted
+)
```
Is endorsed not supported in this API?

peter-fogg replied:
Endorsed is not supported.

cahrens commented on lms/djangoapps/discussion_api/tests/test_api.py:
```diff
@@ -1570,7 +1580,14 @@ def test_invalid_field(self):
 @ddt.ddt
-class CreateCommentTest(CommentsServiceMockMixin, UrlResetMixin, SharedModuleStoreTestCase):
+@disable_signal(api, 'comment_created')
+@disable_signal(api, 'comment_voted')
```
Again, coverage for voted.

cahrens commented on lms/djangoapps/discussion_api/tests/test_views.py:
```diff
@@ -391,6 +393,7 @@ def setUp(self):
         super(ThreadViewSetCreateTest, self).setUp()
         self.url = reverse("thread-list")
 
+    @disable_signal(api, 'thread_created')
```
I think I would put the disabling on the class, like you did in the previous test. The reason being that more test cases might be added to the tests, and it might not be clear to the author that they should disable the signals.

peter-fogg replied:
Done.

cahrens commented on lms/djangoapps/teams/models.py:
```diff
@@ -1,18 +1,74 @@
+@receiver(thread_deleted)
+@receiver(comment_edited)
+@receiver(comment_deleted)
+def post_edit_delete_handler(sender, **kwargs):  # pylint: disable=unused-argument
```
Can you create a helper method for these 3 repeated blocks of code (with a boolean to indicate whether the user who performed the action must also be original author?

peter-fogg replied:
Done.

cahrens commented on lms/djangoapps/teams/tests/test_models.py:
```diff
@@ -73,3 +89,79 @@ def test_user_in_team_for_course(self, username, course_id, expected_value):
+        """
+        team_last_activity = self.team.last_activity_at
+        team_membership_last_activity = self.team_membership.last_activity_at
+        user = getattr(self, user)
```
Is it also worth testing voting on someone else's post (assuming that both users are in the team)? That should update the last_activity date.

peter-fogg replied:
Done.

cahrens commented on lms/lib/comment_client/comment.py:
```diff
@@ -30,6 +30,11 @@ class Comment(models.Model):
         return Thread(id=self.thread_id, type='thread')
 
+    @property
+    def context(self):
```
You don't use this property now that you are doing getattr, correct? Any reason to keep this code?

peter-fogg replied:
This is still useful to allow the signal handlers to work for both threads and comments. If it were removed, the `getattr(post, 'context', 'course')` call would always return `'course'` for comments.

cahrens commented:
A couple of small comments, but otherwise :+1: This is looking good!

peter-fogg commented:
jenkins run all?

dianakhuang commented on common/test/utils.py:
```diff
@@ -0,0 +1,77 @@
+
+def disable_signal(module, signal):
+    """Replace `signal` inside of `module` with a dummy signal. Can be
+    used as a method or class decorator, as well as a context manager."""
```
This explanation seems at odds with the name. We're really just mocking out the signal, not actually disabling it, right?

peter-fogg replied:
Well, the intention is for the signal to have no side effects during the test. Since we're not making any assertions about the mock, like if it was called, I think it makes sense to call it "disable".

dianakhuang commented:
Other than that nit, :ship:

cahrens commented:
Need to rebase this, and remember not to merge until we have done the load testing!
